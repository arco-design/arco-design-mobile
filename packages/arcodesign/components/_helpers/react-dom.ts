import { ReactElement } from 'react';
import * as ReactDOM from 'react-dom';

function isObject(obj: any): obj is { [key: string]: any } {
    return Object.prototype.toString.call(obj) === '[object Object]';
}

export interface RootType {
    render: (element: ReactElement) => void;
    _unmount: () => void;
}
export interface RootTypeReact extends RootType {
    unmount?: () => void;
}
export type CreateRootFnType = (container: Element | DocumentFragment) => {
    render: (element: ReactElement) => void;
    unmount?: () => void;
};

// Cast ReactDOM to a version that might have createRoot and other properties
const typedReactDOM = ReactDOM as typeof ReactDOM & {
    createRoot?: CreateRootFnType;
    render?: (
        element: ReactElement,
        container: Element | DocumentFragment,
        callback?: () => void,
    ) => any;
    unmountComponentAtNode?: (container: Element | DocumentFragment) => boolean;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
};

let copyRender: (
    app: ReactElement,
    container: Element | DocumentFragment,
    createRootFunction?: CreateRootFnType,
) => RootType;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
    // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
    // Avoid console warning
    const secretInternals = typedReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    if (isObject(secretInternals)) {
        secretInternals.usingClientEntryPoint = skipWarning;
    }
};

const createRootFn = typedReactDOM.createRoot;

const getRender =
    (createRootFunction: CreateRootFnType) =>
    (app: ReactElement, container: Element | DocumentFragment) => {
        updateUsingClientEntryPoint(true);
        const root = createRootFunction!(container);
        updateUsingClientEntryPoint(false);

        root.render(app);

        return {
            render: (elementToRender: ReactElement) => {
                root.render(elementToRender);
            },
            _unmount() {
                setTimeout(() => {
                    if (root && typeof root.unmount === 'function') {
                        root.unmount();
                    }
                });
            },
        };
    };

if (createRootFn) {
    // React 18
    copyRender = getRender(createRootFn);
} else if (
    typeof typedReactDOM.render === 'function' &&
    typeof typedReactDOM.unmountComponentAtNode === 'function'
) {
    // React 16/17
    copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
        typedReactDOM.render!(app, container); // Use non-null assertion
        return {
            render: (elementToRender: ReactElement) => {
                typedReactDOM.render!(elementToRender, container); // Use non-null assertion
            },
            _unmount() {
                typedReactDOM.unmountComponentAtNode!(container); // Use non-null assertion
            },
        };
    };
} else {
    copyRender = (
        app: ReactElement,
        container: Element | DocumentFragment,
        createRootFunction?: CreateRootFnType,
    ) => {
        const defaultCb = () => {
            // Fallback if no rendering method is found
            console.error(
                'ArcoDesign: ReactDOM.createRoot() or ReactDOM.render() not found. ' +
                    'This usually means you are using an unsupported version of React, ' +
                    'or ReactDOM is not properly initialized. ' +
                    'ArcoDesign Mobile React requires React 16, 17, 18, or 19.',
            );
            return { render: (_element: ReactElement) => {}, _unmount: () => {} };
        };
        if (createRootFunction) {
            return getRender(createRootFunction)(app, container);
        }
        return defaultCb();
    };
}

export const render = copyRender;
