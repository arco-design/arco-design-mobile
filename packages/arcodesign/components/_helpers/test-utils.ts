import React, { useEffect, useRef } from 'react';
import { Promise as ES6Promise, Promise } from 'es6-promise';

/**
 * React 19 兼容的 act 包装器
 * 用于包装可能触发状态更新的异步操作
 */
export const actWrapper = (callback: () => void | ES6Promise<void>): ES6Promise<void> => {
    return new ES6Promise<void>((resolve, reject) => {
        try {
            // 尝试使用 react-dom/test-utils 的 act
            const testUtils = require('react-dom/test-utils');
            if (testUtils && typeof testUtils.act === 'function') {
                // 对于同步回调
                if (typeof callback === 'function') {
                    try {
                        const result = callback();
                        if (result && typeof result.then === 'function') {
                            testUtils
                                .act(() => result)
                                .then(() => resolve())
                                .catch(reject);
                        } else {
                            testUtils.act(() => {
                                callback();
                            });
                            resolve();
                        }
                    } catch (error) {
                        reject(error);
                    }
                } else {
                    resolve();
                }
            } else {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            }
        } catch (error) {
            // 如果 act 不可用，直接执行回调
            try {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            } catch (callbackError) {
                reject(callbackError);
            }
        }
    });
};

/**
 * 清理未处理的异步操作，防止内存泄漏警告
 */
export const cleanupAsync = (): ES6Promise<void> => {
    return new ES6Promise<void>(resolve => {
        setTimeout(() => resolve(), 0);
    });
};

/**
 * 安全的状态更新，检查组件是否已卸载
 */
export const createSafeStateUpdater = (isMountedRef: React.MutableRefObject<boolean>) => {
    return (updater: () => void) => {
        if (isMountedRef.current) {
            updater();
        }
    };
};

/**
 * 用于测试的 useEffect 清理函数
 */
export const useTestCleanup = (cleanupFn: () => void) => {
    const isMountedRef = useRef(true);

    useEffect(() => {
        return () => {
            isMountedRef.current = false;
            cleanupFn();
        };
    }, [cleanupFn]);

    return isMountedRef;
};
