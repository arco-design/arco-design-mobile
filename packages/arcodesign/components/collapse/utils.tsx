import React, { Context } from 'react';
import { GroupContextParams } from './type';

export const defaultContextKey = '__COLLAPSE_CONTEXT__';
export const allContexts: Record<string, Context<GroupContextParams>> = {};

export function getCollapseContext(key: string) {
    if (!allContexts[key]) {
        allContexts[key] = React.createContext<GroupContextParams>({
            isGroup: false,
            value: [],
            onCollapse: () => {},
        });
    }
    return allContexts[key];
}

export const CollapseKeyContext = React.createContext<{ key: string }>({
    key: defaultContextKey,
});
