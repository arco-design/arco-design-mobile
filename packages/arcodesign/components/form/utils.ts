import { IRules, ValidatorError } from '@arco-design/mobile-utils';

export const isFieldRequired = (rules: IRules[] = []) => {
    return (rules || []).some(rule => rule?.required);
};

export const getErrorAndWarnings = (result: ValidatorError[]) => {
    let errors: string[] = [];
    let warnings: string[] = [];
    result.map(({ message = [], validateLevel = 'error' }) => {
        if (!message?.length) {
            return;
        }
        if (validateLevel === 'warning') {
            warnings = [...warnings, ...message];
        } else {
            errors = [...errors, ...message];
        }
    });
    return { warnings, errors };
};
