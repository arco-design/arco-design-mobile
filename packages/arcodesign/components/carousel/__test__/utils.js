import { act } from 'react-dom/test-utils';
import { createMoveTouchEventObject, createStartTouchEventObject } from '../../../tests/helpers/mockEvent';

export function mockSwipeStart(map, clientXMap) {
    const { touchstart, touchmove } = clientXMap;
    act(() => {
        map.touchstart(createStartTouchEventObject({ x: touchstart, y: 0 }));
        map.touchmove(createMoveTouchEventObject({ x: touchmove, y: 0 }));
    });
}
export function mockSwipeEnd(wrapper, className, touchend) {
    wrapper.find(`.${className}`).simulate('touchEnd', createMoveTouchEventObject({ x: touchend, y: 0 }));
    act(() => {
        jest.advanceTimersByTime(600);
    });
    wrapper.update();
}

export function mockSwipe(map, wrapper, className, clientXMap) {
    const { touchend } = clientXMap;
    mockSwipeStart(map, clientXMap);
    mockSwipeEnd(wrapper, className, touchend);
}
