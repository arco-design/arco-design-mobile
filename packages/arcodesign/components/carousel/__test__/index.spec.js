import React from 'react';
import { render, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import Carousel from '..';
import ContextProvider, { defaultContext } from '../../context-provider';
import { createMoveTouchEventObject, mockAddListener, mockDocumentVisibility } from '../../../tests/helpers/mockEvent';
import { mockContainerSize, resetContainerSizeMock } from '../../../tests/helpers/mockElement';
import { mockSwipe } from './utils';

demoTest('carousel');

mountTest(Carousel, 'Carousel');

const prefix = `${defaultContext.prefixCls}-carousel`;

function indexIsActive(wrapper, index) {
    const items = wrapper.find(`.${prefix}-item.normal-item`);
    return items.at(index).hasClass('active');
}

describe('Carousel', () => {

    beforeAll(() => {
        mockContainerSize();
    });

    afterAll(() => {
        resetContainerSizeMock();
    });

    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.useRealTimers();
    });

    it('should auto play correctly', () => {
        const onChange = jest.fn();
        const wrapper = mount(<Carousel stayDuration={1000} boxWidth={375} onChange={onChange}>
            <div />
            <div />
        </Carousel>);
        act(() => {
            jest.advanceTimersByTime(1600);
        });
        wrapper.update();
        expect(indexIsActive(wrapper, 1)).toBe(true);
        expect(onChange.mock.calls).toHaveLength(1);
        expect(onChange.mock.calls[0]).toEqual([1]);
    });

    it('should render correctly and disable loop when only one child', () => {
        const wrapper = mount(<Carousel stayDuration={1000} boxWidth={375}>
            <div />
        </Carousel>);
        const inner = wrapper.find(`.${prefix}-inner`);
        expect(inner.children).toHaveLength(1);
        act(() => {
            jest.advanceTimersByTime(1600);
        });
        wrapper.update();
        expect(indexIsActive(wrapper, 0)).toBe(true);
    });

    it('should render correctly when using the property `list`', () => {
        const wrapper = render(<Carousel list={[
            { src: '11', text: 'test 1' },
            { src: '22', text: 'test 2' },
        ]} />);
        const items = wrapper.find(`.${prefix}-item`);
        expect(items).toHaveLength(2);
    });

    it('should disable loop playback when using the property `offsetBetween`', () => {
        const wrapper = mount(<Carousel stayDuration={1000} boxWidth={375} offsetBetween={20}>
            <div />
            <div />
        </Carousel>);
        act(() => {
            jest.advanceTimersByTime(1600);
        });
        wrapper.update();
        expect(indexIsActive(wrapper, 0)).toBe(true);
    });

    it('should enable loop playback and render fake dom when using the property `offsetBetween` and `fakeItem`', () => {
        const wrapper = mount(<Carousel
            stayDuration={1000}
            boxWidth={375}
            offsetBetween={20}
            fakeItem={true}
            initialIndex={2}
        >
            <div />
            <div />
            <div />
        </Carousel>);
        act(() => {
            jest.advanceTimersByTime(1600);
        });
        wrapper.update();
        expect(indexIsActive(wrapper, 0)).toBe(true);
        const items = wrapper.find(`.${prefix}-item.fake-item`);
        expect(items).toHaveLength(3);
    });

    it('should support using ref to change index', () => {
        const ref = React.createRef();
        const wrapper = mount(<Carousel autoPlay={false} boxWidth={375} ref={ref}>
            <div />
            <div />
        </Carousel>);
        const { changeIndex } = ref.current;
        expect(typeof changeIndex).toBe('function');
        act(() => {
            changeIndex(1);
            jest.advanceTimersByTime(600);
        });
        wrapper.update();
        expect(indexIsActive(wrapper, 1)).toBe(true);
    });

    it('should handle touch event correctly', () => {
        const wrapper = mount(<Carousel
            autoPlay={false}
            boxWidth={375}
            swipeable={false}
        >
            <div />
            <div />
        </Carousel>);
        const map = mockAddListener(wrapper.find(`.${prefix}`));
        wrapper.setProps({ swipeable: true });

        // move to last item
        mockSwipe(map, wrapper, prefix, { touchstart: 100, touchmove: 150, touchend: 300 });
        expect(indexIsActive(wrapper, 1)).toBe(true);

        // move to next item
        mockSwipe(map, wrapper, prefix, { touchstart: 300, touchmove: 150, touchend: 100 });
        expect(indexIsActive(wrapper, 0)).toBe(true);

        // move and cancel
        wrapper.setProps({ speedToChange: Infinity });
        mockSwipe(map, wrapper, prefix, { touchstart: 100, touchmove: 150, touchend: 100 });
        expect(indexIsActive(wrapper, 0)).toBe(true);
    });

    it('should support using `onTouchStart` to prevent default logic', () => {
        const onTouchStart = jest.fn(() => true);
        const onTouchEnd = jest.fn(() => true);
        const wrapper = mount(<Carousel
            autoPlay={false}
            boxWidth={375}
            onTouchStart={onTouchStart}
            onTouchEnd={onTouchEnd}
            swipeable={false}
            autoHeight={true}
        >
            <div />
            <div />
        </Carousel>);
        // should not trigger touchend when swipeable=false
        wrapper.find(`.${prefix}`).simulate('touchEnd', createMoveTouchEventObject({ x: 100, y: 0 }));
        expect(onTouchEnd.mock.calls).toHaveLength(0);

        // should prevent default logic when onTouchStart or onTouchEnd return true
        const map = mockAddListener(wrapper.find(`.${prefix}`));
        wrapper.setProps({ swipeable: true });
        mockSwipe(map, wrapper, prefix, { touchstart: 100, touchmove: 150, touchend: 300 });
        expect(onTouchStart.mock.calls).toHaveLength(1);
        expect(indexIsActive(wrapper, 0)).toBe(true);
    });

    it('should hide carousel when invisible and rerender forcibly when visible in ios', () => {
        const wrapper = mount(<ContextProvider system="ios">
            <Carousel boxWidth={375}>
                <div />
                <div />
            </Carousel>
        </ContextProvider>);
        const dom = wrapper.find(`.${prefix}:not(.wrap-placeholder)`).at(0).getDOMNode();
        act(() => {
            mockDocumentVisibility('hidden');
        });
        expect(getComputedStyle(dom).getPropertyValue('display')).toBe('none');
        act(() => {
            mockDocumentVisibility('visible');
        });
        expect(getComputedStyle(dom).getPropertyValue('display')).toBe('block');
    });
});
