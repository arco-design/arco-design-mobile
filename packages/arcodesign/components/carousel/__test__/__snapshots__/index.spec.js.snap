// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`carousel demo test carousel demo: basic.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-horizontal"
    >
      <i
        class="indicator type-circle active"
      />
      <i
        class="indicator type-circle"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: change.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_2.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-horizontal"
    >
      <i
        class="indicator type-square active"
      />
      <i
        class="indicator type-square"
      />
      <i
        class="indicator type-square"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: change-index.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-horizontal"
    >
      <i
        class="indicator type-circle active"
      />
      <i
        class="indicator type-circle"
      />
    </div>
  </div>
  <div
    class="change-index-btns"
  >
    <button
      class="arco-button arco-button-type-primary type-primary arco-button-size-small arco-button-size-small-is-semi size-small is-semi pc"
      type="button"
    >
      <div
        class="arco-button-text arco-button-text-pc btn-text"
      >
        Previous
      </div>
    </button>
    <button
      class="arco-button arco-button-type-primary type-primary arco-button-size-small arco-button-size-small-is-semi size-small is-semi pc"
      type="button"
    >
      <div
        class="arco-button-text arco-button-text-pc btn-text"
      >
        Next
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: custom-width.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: 900px; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 300px; padding-left: 0px; box-sizing: content-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item"
          style="transform: translateX(0px); width: 300px; padding-left: 0px; box-sizing: content-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item"
          style="transform: translateX(0px); width: 300px; padding-left: 0px; box-sizing: content-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_2.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-horizontal"
    >
      <i
        class="indicator type-square active"
      />
      <i
        class="indicator type-square"
      />
      <i
        class="indicator type-square"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: indicator.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_2.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-end arco-carousel-indicator-horizontal"
    >
      <span
        class="radius-num-indicator"
      >
        1/3
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: lazy.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        />
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        />
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-horizontal"
    >
      <i
        class="indicator type-circle active"
      />
      <i
        class="indicator type-circle"
      />
      <i
        class="indicator type-circle"
      />
      <i
        class="indicator type-circle"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: list.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            class="carousel-item-img needsclick"
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
          <div
            class="carousel-item-text"
          >
            First carousel
          </div>
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            class="carousel-item-img needsclick"
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_2.jpg"
          />
          <div
            class="carousel-item-text"
          >
            Second carousel
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-end arco-carousel-indicator-horizontal"
    >
      <i
        class="indicator type-circle active"
      />
      <i
        class="indicator type-circle"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: offset-space.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(-100%) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(65px); width: 100%; padding-left: 35px; padding-right: 35px; box-sizing: border-box;"
        >
          <div
            class="arco-carousel-item-inner"
            style="transition-property: none; transition-duration: 0ms; transform: translateX(0%) scale(0.8); transform-origin: 100% center;"
          >
            <img
              alt=""
              src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
            />
          </div>
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 35px; padding-right: 35px; box-sizing: border-box;"
        >
          <div
            class="arco-carousel-item-inner"
            style="transition-property: none; transition-duration: 0ms; transform: translateX(0%) scale(1); transform-origin: 100% center;"
          >
            <img
              alt=""
              src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
            />
          </div>
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(-65px); width: 100%; padding-left: 35px; padding-right: 35px; box-sizing: border-box;"
        >
          <div
            class="arco-carousel-item-inner"
            style="transition-property: none; transition-duration: 0ms; transform: translateX(0%) scale(0.8); transform-origin: 0% center;"
          >
            <img
              alt=""
              src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_2.jpg"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-horizontal"
    >
      <i
        class="indicator type-square"
      />
      <i
        class="indicator type-square active"
      />
      <i
        class="indicator type-square"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: outside.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
    >
      <div
        class="arco-carousel-inner auto"
        style="width: auto; transition-property: none; transition-duration: 0ms; transform: translateX(0) translateZ(0);"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item ssr-float"
          style="transform: translateX(0px); width: 100%; padding-left: 0px; padding-right: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_2.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-horizontal inverse outside"
    >
      <i
        class="indicator type-circle active"
      />
      <i
        class="indicator type-circle"
      />
      <i
        class="indicator type-circle"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: vertical.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
      style="height: 140px;"
    >
      <div
        class="arco-carousel-inner auto vertical"
        style="transition-property: none; transition-duration: 0ms; transform: translateY(0) translateZ(0); width: 100%;"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active vertical"
          style="transform: translateY(0px); height: 140px; padding-top: 0px; padding-bottom: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_5.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item vertical ssr-float"
          style="transform: translateY(0px); height: 140px; padding-top: 0px; padding-bottom: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_1.jpg"
          />
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item vertical ssr-float"
          style="transform: translateY(0px); height: 140px; padding-top: 0px; padding-bottom: 0px; box-sizing: border-box;"
        >
          <img
            alt=""
            src="https://sf1-cdn-tos.toutiaostatic.com/obj/arco-mobile/_static_/large_image_2.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="arco-carousel-indicator pos-center arco-carousel-indicator-vertical vertical ver-pos-left"
    >
      <i
        class="indicator type-square active"
      />
      <i
        class="indicator type-square"
      />
      <i
        class="indicator type-square"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`carousel demo test carousel demo: vertical-text.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-carousel-wrap"
  >
    <div
      class="arco-carousel"
      style="height: 22px;"
    >
      <div
        class="arco-carousel-inner auto vertical"
        style="transition-property: none; transition-duration: 0ms; transform: translateY(0) translateZ(0); width: 100%;"
      >
        <div
          class="arco-carousel-item carousel-item normal-item active vertical"
          style="transform: translateY(0px); height: 22px; padding-top: 0px; padding-bottom: 0px; box-sizing: border-box;"
        >
          <div
            class="vertical-text"
          >
            Carousel 1
          </div>
        </div>
        <div
          class="arco-carousel-item carousel-item normal-item vertical ssr-float"
          style="transform: translateY(0px); height: 22px; padding-top: 0px; padding-bottom: 0px; box-sizing: border-box;"
        >
          <div
            class="vertical-text"
          >
            Carousel 2
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
